### PR TITLE
Create contrib module

### PR DESCRIPTION
**Pull request by olafurpg**
cc/ #537 @DavidDudson 

Contains:

- Equal typeclass for type-safe comparison
- Collection like operations on trees
- Utility to find leading/trailing comments

### CODE REVIEW

File: scalameta/contrib/src/main/scala/scala/meta/contrib/ContribApi.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    def contains[F[_ <: Tree]](toFind: Tree)(implicit conv: Tree => F[Tree],
+                                             eqEv: Equal[F[Tree]]): Boolean =
+      find(_.equal[F](toFind)).nonEmpty
+  }
+
+}
+
+object TreeOps {
+  @tailrec
+  final def parents(tree: Tree, accum: Seq[Tree] = Seq.empty[Tree]): Seq[Tree] = {
```
Comment (DavidDudson) [inline scalameta/contrib/src/main/scala/scala/meta/contrib/ContribApi.scala]: Can we name this ancestors instead. And the have the opposite too.

- parent == this tree's parent
- children == this tree's children
- descendants = this tree's children, and grandchildren etc.
- ancestors = this trees parents, and grandparents etc.

Comment (DavidDudson) [inline scalameta/contrib/src/main/scala/scala/meta/contrib/ContribApi.scala]: Also it seems weird that this is in a treeops object, whereas `collectFirst` is not, where `collectFirst` is longer

Comment (olafurpg) [inline scalameta/contrib/src/main/scala/scala/meta/contrib/ContribApi.scala]: Good points.

1. Sure.
2. I was thinking of putting all the methods in an object and make the pimps optional. The pimps could for example be added with a `syntax` import, as is convention in other libs.

Comment (DavidDudson) [inline scalameta/contrib/src/main/scala/scala/meta/contrib/ContribApi.scala]: Actually, can you hold up on modifying these. I've been extending this PR and it makes alot of my comments redundant. I'm about to push

Comment (olafurpg) [inline scalameta/contrib/src/main/scala/scala/meta/contrib/ContribApi.scala]: Sorryy, just pushed a small update. I look forward to see your changes tomorrow, I'm hitting bed now :)

Comment (DavidDudson) [inline scalameta/contrib/src/main/scala/scala/meta/contrib/ContribApi.scala]: https://github.com/DavidDudson/scalameta/commit/80aa21b6fb5d725a5107481666616b1a396953b6

Depending on your opinions on this, I may create a PR for your PR. Or else i can keep it seperate

**Comment by xeno-by** [comment]
I agree that there's a big need in a place where we could put a grab bag of utilities without polluting the main API. Once certain utilities get popular, we can consider graduating them into scalameta/scalameta. If they don't get popular, they can stay in scalameta/contrib without confusing the newcomers.

I'm totally up for merging this, apart from one comment. I suggest that we follow the convention that's already in place in the main library and don't require extra imports apart from `import scala.meta.contrib._`.

File: scalameta/contrib/src/main/scala/scala/meta/contrib/ContribApi.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
+package scala.meta.contrib
+
+import scala.language.higherKinds
+import scala.language.implicitConversions
+
+import scala.annotation.tailrec
+import scala.collection.immutable.Seq
+import scala.meta._
+
+trait ContribApi {
```
Comment (xeno-by) [inline scalameta/contrib/src/main/scala/scala/meta/contrib/ContribApi.scala]: How about merging this trait into `package.scala`?

**Comment by olafurpg** [comment]
I updated the PR to include more collections operations from @DavidDudson's commit (took the liberty to include you as an author to the commit). I'm happy with the PR as it is now, feel free to merge @xeno-by.

**Comment by DavidDudson** [comment]
@olafurpg LGTM

**Comment by xeno-by** [comment]
LGTM

**Pull request merged**